File: internal/storage/ratelimit.lua
-- Sliding window rate limiter
-- KEYS[1]: rate limit key (e.g., "ratelimit:192.168.1.1")
-- ARGV[1]: window_ms - sliding window size in milliseconds
-- ARGV[2]: limit - max requests allowed in window
-- ARGV[3]: ttl - key expiration in seconds

local key = KEYS[1]
local window_ms = tonumber(ARGV[1])
local limit = tonumber(ARGV[2])
local ttl = tonumber(ARGV[3])

local time_result = redis.call('TIME')
local now = tonumber(time_result[1]) * 1000 + math.floor(tonumber(time_result[2]) / 1000)
local window_start = now - window_ms

redis.call('ZREMRANGEBYSCORE', key, '-inf', window_start)

local count = redis.call('ZCARD', key)

if count < limit then
    local member = tostring(now) .. ':' .. tostring(math.random(1000000))
    redis.call('ZADD', key, now, member)
    redis.call('EXPIRE', key, ttl)
    return 1
else
    return 0
end

File: internal/storage/ratelimit.go
package storage

import (
	"context"
	_ "embed"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

//go:embed ratelimit.lua
var rateLimitLua string

var rateLimitScript = redis.NewScript(rateLimitLua)

type rateLimitParams struct {
	window time.Duration // ARGV[1]: sliding window size in milliseconds
	limit  int           // ARGV[2]: max requests allowed in window
	ttl    time.Duration // ARGV[3]: key expiration in seconds
}

func (p rateLimitParams) args() []any {
	return []any{
		p.window.Milliseconds(),
		p.limit,
		int(p.ttl.Seconds()),
	}
}

func runRateLimitScript(ctx context.Context, client *redis.Client, key string, params rateLimitParams) (bool, error) {
	result, err := rateLimitScript.Run(ctx, client,
		[]string{key},
		params.args()...,
	).Int()
	if err != nil {
		return false, fmt.Errorf("failed to run rate limit script: %w", err)
	}
	return result == 1, nil
}
